### PR TITLE
Implemented methods required by tests for UserServiceImpl

### PR DESCRIPTION
**Pull request by Valiusikk**

### CODE REVIEW

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 1]
```diff
@@ -2,68 +2,99 @@
...
+                map(User::getFirstName).
+                sorted(Comparator.reverseOrder()).
+                collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                sorted(Comparator.comparing(User::getFirstName)).
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: you can use construction like 
`comparing(age).reversed().thenComparing(name)`

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 2]
```diff
@@ -2,68 +2,99 @@
...
+                sorted(Comparator.comparing(User::getAge).reversed())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                map(User::getPrivileges).
+                flatMap(List::stream).
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: please add also distinct here

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 3]
```diff
@@ -2,68 +2,99 @@
...
+                collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                filter(user -> user.getAge() > age && user.getPrivileges().contains(Privilege.UPDATE))
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: Why do we need to check privilege here

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 4]
```diff
@@ -2,68 +2,99 @@
...
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                map(User::getAge).
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: `mapToInt(User::getAge)`
`mapToInt` do Integer::valueOf or Integer::intValue

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 5]
```diff
@@ -2,68 +2,99 @@
...
+                average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().
+                collect(Collectors.groupingBy(User::getLastName, Collectors.counting())).//group by lastname and count how many times it repeats
+                        entrySet().stream().filter(node -> node.getValue() >= 2).//put condition that count should be more than 1
+                        reduce((a, b) -> a.getValue() < b.getValue() ? b ://look for node with biggest count
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: You can use max

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java [hunk 6]
```diff
@@ -26,7 +26,7 @@
     private static final List<Privilege> ALL_PRIVILEGES = asList(Privilege.values());
 
     @Test
-    void shouldReturnFirstNamesSortedDesc() {
+    public void shouldReturnFirstNamesSortedDesc() {
```
Comment (Evgenikaall) [inline src/test/java/com/endava/internship/service/UserServiceImplTest.java]: why ? For tests, access modifiers must be package-private

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 7]
```diff
@@ -2,68 +2,99 @@
```
Comment (Evgenikaall) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java:2]: for streams, try to keep this formatting :
```
list.stream()
    .map(Util::something)
    .collect(toSet())
```

Comment (Evgenikaall) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java:2]: each stream function must be started with dot

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 3 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: /**
     * Should return optional of user that has UPDATE privilege
     * and age higher that given age
     *
     * @param users list of users
     * @param age   conditional age
     * @return optional of found user or optional empty if user wasn't found
     */
    Optional<User> getUpdateUserWithAgeHigherThan(List<User> users, int age);

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 5 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java (hunk 6 above)
Comment (Valiusikk) [reply src/test/java/com/endava/internship/service/UserServiceImplTest.java]: fixed

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 2 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: fixed

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: fixed

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 7 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java:2]: Fixed!

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 3 above)
Comment (Evgenikaall) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: all good

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 4 above)
Comment (Valiusikk) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: fixed!

**Review by Evgenikaall — APPROVED** [review]

**Pull request merged**